### DISCUSSION TITLE
Blind Luck MUD Development Diary

### DISCUSSION THREAD

**Discussion by Enfors** [evennia/evennia#2569] [My Evennia Project]
### Introduction
I'm making a game that tries to make sure it is as accessible to the visually impaired as it is to the rest of us. I'm calling this game Blind Luck MUD. On the Discord server, I mentioned that I write articles about the development of this MUD, and Griatch asked me to post links to them here.

### Full disclosure
In the interest of full disclosure, I will say that I'm paid to write these articles - but unless this thread for some reason would generate tens of thousands of new views on the articles (which it won't), I will not get paid anything more because I post them here. In other words, I'm not posting here for financial reasons. Instead, I'm hoping to garner some interesting discussions.

### Bypassing Medium's paywall
Medium allows people to read up to three articles for free per month, and in the unlikely event that anyone will run into that limit, please let me know and I'll be able to provide you with a "friend link" which bypasses the paywall.

### The articles published so far
[Blind Luck MUD Development Diary, Entry 1](https://enfors.medium.com/blind-luck-mud-development-diary-entry-1-4ee4f6dc73a2): I'm making an online fantasy RPG that even the visually impaired can play. Will I succeed? Read along, and let's find out together!

[Entry 2](https://enfors.medium.com/blind-luck-mud-developer-diary-entry-2-249148023991): A small start.

[Entry 3: Experience, Levels & Stats](https://enfors.medium.com/blind-luck-mud-development-diary-entry-3-experience-levels-stats-e5168cc53c26): How I implemented the initial version of experience points, the levels system, and stats.

[Entry 4: Rooms](https://enfors.medium.com/blind-luck-mud-development-diary-entry-4-rooms-991c730de946): My approach to creating areas and rooms. If you ask me, this is one of the hardest parts of making a MUD even if it's not a technically difficult task.

[Entry 5: Gender in games](https://enfors.medium.com/blind-luck-mud-development-diary-entry-4-rooms-991c730de946): Do characters in games really need a gender, or are preferred pronouns enough?

[Entry 6: Blind mode](https://enfors.medium.com/blind-luck-mud-development-diary-entry-6-blind-mode-2a0f12871567): It's about time Blind Luck MUD started living up to its name and added features to help the visually impaired.

[Entry 7: Combat planning](https://enfors.medium.com/blind-luck-mud-development-diary-entry-7-combat-planning-f860f3e6e412): Combat will be a core mechanic in Blind Luck MUD. Getting it right is important, which is part of the reason why I'm writing an article about it. Another part of the reason why I'm writing an article only about planning may or may not be because it turns out implementing this stuff is taking longer than expected - I will never tell.

**Comment by Griatch** [comment]
Some interesting reading here. I like your approach of describing your game as a series of growing cicles, where each circle contains all the basic pieces of the game but in a simple form. ![Expanding circle](https://miro.medium.com/max/760/1*DaCU-IBmpDToSZRRtOzTxA.png)
Seems like a pretty natural way to make sure to reach intermediary goals while having something 'playable' in every step.

**Reply by Enfors** [reply]
Thank you! Yes, that's the idea. It's sort of like a one-man Agile approach, I guess? Get *something* working as soon as possible, then slowly expand on it over time.

**Reply by Griatch** [reply]
The Evennia documentation suggests to "start by making a small tech-demo first". You take this even further by divvying things up in even smaller steps with this 'circle approach'. 
I can only read up to the third article on Medium, have you been able to keep up following this growing-circle method as the game's systems become more complex? I could for example imagine a lot of overlap between Character and Combat sections and the world-part of the circle lagging behind as you focus on one side of the diagram.

**Reply by Enfors** [reply]
Well, mostly in the later articles I haven't been using the circle graphic because I feel I haven't added enough to warrant another slot being filled in. Once I get a basic version of the combat system working, then I'll fill in another slot there.

In case you want to read any of the later articles, here are "friend links" which bypass the paywall:

[Part 3: Experience, Levels & Stats](https://enfors.medium.com/blind-luck-mud-development-diary-entry-3-experience-levels-stats-e5168cc53c26?sk=d0c5049c8932746518743136d9091c1e)

[Part 4: Rooms](https://enfors.medium.com/blind-luck-mud-development-diary-entry-4-rooms-991c730de946?sk=e581a4feac2111223d21cdd6b4c471f7)

[Part 5: Gender in games](https://enfors.medium.com/blind-luck-mud-developer-diary-entry-5-gender-in-games-3642dffaadd5?sk=5e0cf6dcf85889a7e866cc05f8e379a5)

[Part 6: Blind mode](https://enfors.medium.com/blind-luck-mud-development-diary-entry-6-blind-mode-2a0f12871567?sk=8e282fc5ef01efec277d096c32e3bb31)

[Part 7: Combat planning](https://enfors.medium.com/blind-luck-mud-development-diary-entry-7-combat-planning-f860f3e6e412?sk=7a3ad98f83f5c59a84916e42081d07c6)

**Comment by Griatch** [comment]
Thanks for the links! I just read the `Blind mode` article and was intrigued by this. I didn't see you mention Evennia's native `screenreader-mode` though, which is used to strip 'ascii-graphics' (for any game) to make it easier to use with screenreaders. Your 'blind mode' is more specific for sure, just letting you know Evennia does have something like this out of the box. ;)

**Reply by Enfors** [reply]
You're right, I should probably have mentioned Evennia's built-in screenreader mode. I added blind mode because I wanted more than just stripping out certain characters. For example, on the MUDs I've played you typically get a "hp bar" displayed during each combat round and when you type "hp" which looks something like this:
```
HP: 37/42  SP: 14/56  Enemy: bleeding
```
I'd imagine this being a bit confusing listening to on a screenreader, so in case the user has blind mode on, I'd prefer if the MUD would instead output something like:
```
Hit points: 37 out of 42. Spell points: 14 out of 56. Your enemy is bleeding.
```
I'm sure other examples of where this is useful will pop up as I implement more and more of the MUD. I suppose you could say that in general, you'd want the "blind" messages to read more like a sentence for the benefit of the screenreader, and less like a "status line" or dump of data.

**Reply by Griatch** [reply]
No problem, just wanted to make sure you were aware of the native screenreader mode since supporting screen readers is such a heavy focus of your game. :) 

As for the brevity, I recall being surprised at how quickly blind/visually impaired people parse text with modern tools. Screen readers often run at many times faster speed than spoken text. And there are people using braille-readers too - they almost read an entire line at a time. 

I'm sure your careful thought about how to present the health bar etc is very appreciated. But don't be surprised if some blind users may prefer the ability to get the shorter presentation instead.
 
If would be great to hear the input of @vincent-lg, if he's still around - he has contributed a lot of code to Evennia (and is a very fast coder too, using a braille reader). I'm sure he'd have a lot of insight to provide.

**Reply by P-Fitzgerald** [reply]
Perhaps a way to configure the way output is reported so the user can choose. This way you could report percentages instead of numbers as if mudflation takes effect those numbers could potentially be very high (IE. Your screen reader could be reporting 'Three Thousand Seven Hundred and Thirty Six out of Four Thousand Two Hundred and Twenty Five' which is far more cumbersome than reporting '88 percent'

**Reply by Enfors** [reply]
Hmm, that's a good point @P-Fitzgerald. Even with lower numbers, "54 percent" is a lot shorter than "eighty-four out of one hundred and ten". I'll have to take that into consideration as well - it could easily be implemented with my `blind_msg` function:
```
character.blind_msg(f"{hp}/{max_hp}",     # Non-blind version
                    f"{hp_percentage}%")  # Blind version

**Reply by Enfors** [reply]
> As for the brevity, I recall being surprised at how quickly blind/visually impaired people parse text with modern tools. Screen readers often run at many times faster speed than spoken text. And there are people using braille-readers too - they almost read an entire line at a time.

Yes, that's true. I must be careful not to underestimate people - I don't want my game to condescend to anyone, just make sensible adjustments such as having combat being turn-based rather than real-time. Even if screenreaders and their users are fast enough to keep up with "You hit the enemy hard. You hit the enemy very hard. The enemy missed you." etc every two seconds, that doesn't mean it's a good experience for them.

**Reply by Griatch** [reply]
> ...such as having combat being turn-based rather than real-time. Even if screenreaders and their users are fast enough to keep up with "You hit the enemy hard. You hit the enemy very hard. The enemy missed you." etc every two seconds, that doesn't mean it's a good experience for them.

As someone who doesn't have a visual impairment and _still_ prefer turn-based combat in order to keep up in a text-game, I agree with this. :wink:

**Comment by vincent-lg** [comment]
Hi everyone,
The subject captured my attention.  I have, unfortunately, not a lot of time these days, even for personal projects, but someone's "interesting project" is still worth a comment and some help, if I can provide them.
First and foremost, congratulations for this project!  I do hope it becomes a successful project.  Accessibility, even in text-based games, is not considered all the time and I find myself talking a lot with MUD administrators to figure out how they could improve a project for screen readers, a project that perhaps was built several years (or decades) ago.
As @Griatch pointed out, not all screen reader users rely on the same tool: to over-simplify, a screen reader allows to gather data about the screen (current application, controls and so on).  The choice of output is not that straightforward.  Most screen readers users rely on a synthesizer to get text.  And yes, most of these users have set the speech rate so high that they can read pretty quickly incoming text.  I don't rely on a synthesizer myself.  My screen reader is turned to "mute" most of the time, I simply rely on a Braille display.  I don't think I'm such a fast reader and I don't think I'm as fast as anyone else reading a text, but that's the output I rely on.  Some users have them both.  Some might have things I don't know about (hey, I can't know it all).  For all these reasons, it's complicated to come up with a single solution that will come in handy in every situation.  I've talked a lot about the screenreader flag in Evennia.  I understand why it does what it does and I haven't suggested to remove it or alter it.  But I know in most Evennia-based games I turn it off, because it does something that doesn't really help me: I enjoy a table.  I can't read all form of ASCII arts, but I can read an ASCII map (most of the time anyway) though it might take me some time.  With a Braille display it's a matter of memorizing, but it's not impossible to do.  And I know some users of a synthesizer who have found the way to read ASCII maps too.  I'm not sure how it's done, but they can do it.  Of course, most screen reader users would say that ASCII art doesn't help them.  I personally alter between it and an "all screen reader mode", if it's supported by the MUD, because information can be lost in a special mode.
I recently was talking to administrators about town maps: in most towns of the game, they have maps which describe the layout of shops and trainers and important things.  Most users relying on synthesizers can't access this information.  So they were wondering whether it wouldn't be worthwhile hiding it altogether for screen reader users: I had to say "no, please no" and explain the situation.  The thing is, not all screen reader users can't read these maps.  But even worse than that, hiding the maps would actually give these users less information.  In a town map, it's not such a big deal, but for maps in the wilderness, as can be found on this game, it might be tricky, so I suggested an alternative.  In short, accessibility can be a mess at times.
So I would definitely encourage you to have features that can be individually turned on and off, if possible.  It doesn't mean you can't have an "all on / all off" feature, or even a "Choose the way your screen reader outputs to the screen" feature, for instance, since that might help enabling/disabling some options.  Regarding the health bar for instance, someone suggested percentage: in my opinion, it's a worthwhile question, but even with a Braille reader, a percentage is easier to read than compute big numbers to know how much health you have.  That being said, it might be the case for sighted users too.
I hope all that helps and again, I wish you all the best for this project.  I'm going to read the Medium articles.  Yes, I'd like to write on Medium too, I hope their editor becomes accessible soon... though I'm not so sure it will be that soon.  We'll see!  In the meantime, let's be optimistic!
Good luck,
Vincent-lg

**Reply by Enfors** [reply]
Hmm - thank you very much for your valuable input! You have given me a lot to think about. I'm impressed that you are able to use at least some ASCII maps. And you make a good point about being able to turn off different features of the MUD individually - a blanket "blind mode on or off" isn't granular enough. I will take this into account as I continue to develop the game.

**Comment by kovitikus** [comment]
> ### Full disclosure
> 
> In the interest of full disclosure, I will say that I'm paid to write these articles - but unless this thread for some reason would generate tens of thousands of new views on the articles (which it won't), I will not get paid anything more because I post them here. In other words, I'm not posting here for financial reasons. Instead, I'm hoping to garner some interesting discussions.
> ### Bypassing Medium's paywall
> 
> Medium allows people to read up to three articles for free per month, and in the unlikely event that anyone will run into that limit, please let me know and I'll be able to provide you with a "friend link" which bypasses the paywall.

Why are you using medium if it has a paywall? Is the money you make on these articles really worth potentially turning away readers that would be your core audience? The point is to promote your game and your hard work on these concepts. So why put it behind a paywall for what must amount to a minuscule monetary return?

I'm not asking because I have issues with reading them myself or anything. I just don't understand your logic.

**Reply by Enfors** [reply]
That's absolutely a fair question!

I've been a writer on Medium for a number of years now, and for the past three months, they've had a program going to which writers could apply. If accepted, you have to write at least four articles that meet their quality standards per month. For this, they guarantee the accepted writers a minimum payment of 100 to 500 USD per month. I applied and was accepted at the 200 USD / month level. So then I started considering what I should write about. I've been toying with the idea of making this MUD for a while now, so I thought, "hey, I can make that MUD I've been thinking about and write a sort of development diary about it on Medium." So that's what I did, and that's why they're behind Medium's paywall.

**Reply by kovitikus** [reply]
Interesting that your level of articles only get 200 of 500. I can't imagine what it requires to get to the 500 a month marker. Deep research articles that literally take an entire week to produce?

I guess it's nice to get some extra money from them, if you need some side-income, but honestly it seems like their site is a scam.

They take other people's content, make tons of money on it, then pay the author a fraction of that. I guess their job is to get your article exposure with both their name and standing as a website. But, $500 a month for what could easily amount to a full time job producing a fresh spectacular article per week... that seems like highway robbery as far as compensation goes for hours committed.

A friend and I were reading another article on Medium, unrelated to yours, and he mistook the site as a news organization. I had to inform him that the site is for people posting blogs. He argued that they could at least have the journalistic integrity to mark the article as an opinion piece.

**Reply by Enfors** [reply]
Well, I don't know. I don't think Medium is quite as bad as that. For example, if someone signs up for a membership ($5 per month) through my affiliate link, I get half of that, not like 10%.

**Comment by Enfors** [comment]
### [Blind Luck MUD Development Diary, Entry 8: Dynamic Descriptions](https://enfors.medium.com/blind-luck-mud-development-diary-entry-8-dynamic-descriptions-77e7be925e84)

In this entry, I explain the difference between dynamic vs static descriptions. The initial plan is to add support for having room descriptions include the weather and time of day.

**Reply by Griatch** [reply]
@Enfors I can't read it, but would be interesting to hear what Evennia tools you use for this. :) Are you leveraging the `ExtendedRoom` contrib or going your own route?

**Reply by Enfors** [reply]
Thank you for your interest in this! Here's a link to the story that bypasses the paywall: https://enfors.medium.com/blind-luck-mud-development-diary-entry-8-dynamic-descriptions-77e7be925e84?sk=70c90e1490a093ba207a32c3f71eb7fb

I haven't started implementing any of this yet - the article merely explains what dynamic descriptions are, how I've used them in the past, and what kind of syntax I was thinking of using for them. I have no idea what kind of support Evennia can give me for this, and I might not need it even if there is something that's useful. I guess the only thing I *need* is a way to make Evennia call one of my functions to get the room's description (rather than just grab the contents of a variable if that's how it works by default, I don't know), and I'm set.

**Comment by Griatch** [comment]
@Enfors From reading the article, it sounds like you may be interested in my answer to [this other discussion thread](https://github.com/evennia/evennia/discussions/2565), it explains how you override how an object describes itself.

The `ExtendedRoom` contrib adds support for room descriptions that vary by season (`spring`, `summer`, `autumn` and `winter`) and each seasonal description can have alternate embedded sentences for `morning`, `afternoon`, `evening` and `night`. It does not include weather but it could certainly be extended to do that. It also adds `details` to the room to look at without having to create database objects for each.

[Here is a list of the current contribs](https://github.com/evennia/evennia/blob/master/evennia/contrib/README.md). It's useful to to look into in order to not reinvent the wheel and/or have something to build from. :)

**Comment by Griatch** [comment]
@Enfors Haven't seen any progress here for a while, how's it going? :)

**Reply by Enfors** [reply]
I'm working on other projects at the moment... I'll come back to this one eventually. But I'm glad you asked!
